Document updating configuration via a ConfigMap (#44540)
kubernetes/website: content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md

Reviewer: Sports are fine, but for vehicle brands, let's avoid referring to real companies.

Maybe swap `car` for `fruit`?

Author: I agree. I have changed the example to use ```fruit``` and it is available in the latest commit.

Reviewer 2: @adityasamant25 Great job on creating this new page! I noticed that there's a request in the [issue](https://github.com/kubernetes/website/issues/23101#issue-677612881)  to include instructions on how to make this work with immutable ConfigMaps. I don't see that content in the PR. Is it something I missed or it's not included in the scope of this update?

Author: @dipesh-rawat Thanks for noting it, the scenario for immutable ConfigMaps had slipped my mind. I'll work on including it.

Author: Related issue #42359

---

> title: Updating configuration via a ConfigMap

Reviewer: (tiny nit) page titles should be in title case, according to the [style guide](https://kubernetes.io/docs/contribute/style/style-guide/)

This feedback shouldn't block a merge.

content_type: tutorial
weight: 20
---

<!-- overview -->
This page provides a step-by-step example of updating configuration within a Pod via a ConfigMap and builds upon the [Configure a Pod to Use a ConfigMap](/docs/tasks/configure-pod-container/configure-pod-configmap/) task. At the end of this tutorial, you will understand how to change the configuration for a running application. 

> This tutorial uses the NGINX image as an example.

Reviewer: Why NGINX? We could use something else, because we're not using that image for its usual purpose of being a webserver.

You could change the image to Alpine Linux or Debian and I think it would still work OK. Is that right?

---

Extra credit: add a final step that runs _two_ containers, one writing a file in HTML format to an `emptyDir`, and the other serving the HTML via HTTP. Change the HTML based on a ConfigMap.

Author: I agree. In the latest commit, I have modified the examples to use the ```alpine``` image.

Author: I'll work on the final step of a Pod with two containers, rendering HTML on the basis of a ConfigMap.

Author: In the latest commit, I have added a third section that demonstrates the scenario with two containers.

Create the deployment.
```shell
kubectl apply -f https://k8s.io/examples/deployments/deployment-with-configmap-as-volume.yaml
```

Check the deployment.
```shell
kubectl get deployment configmap-volume
```

You should see an output similar to:
```
NAME               READY   UP-TO-DATE   AVAILABLE   AGE
configmap-volume   3/3     3            3           19s
```

Check the pods.
```shell
kubectl get pods --selector=app=configmap-volume
```

You should see an output similar to:
```
NAME                                READY   STATUS    RESTARTS   AGE
configmap-volume-6b976dfdcf-qxvbm   1/1     Running   0          72s
configmap-volume-6b976dfdcf-skpvm   1/1     Running   0          72s
configmap-volume-6b976dfdcf-tbc6r   1/1     Running   0          72s
```

The data in the ConfigMap is mounted to the Pods as a Volume. The Pod references this data to generate an output that is printed to stdout. Check this by viewing the logs of the Pods.
```shell
kubectl logs <pod-name>
```

You should see the following output printed every 10 seconds:
```
My preferred sport is football
```

Edit the ConfigMap.
```shell
kubectl edit configmap sport
```

> In the yaml editor, change the value of key `sport` from `football` to `cricket` and save.

Reviewer: ```sugggestion
In the editor that appears, change the value of key `sport` from `football` to `cricket. Save your changes.
The kubectl tool updates the ConfigMap accordingly (if you see an error, try again).

Here's an example of how that manifest could look after you edit it:
```

Author: Thanks for this suggestion, it is clearly better. I'll add this in my next commit.

```yaml
apiVersion: v1
data:
  sport: cricket
kind: ConfigMap

> metadata:
>   creationTimestamp: "2023-12-27T08:24:51Z"
>   name: sport
>   namespace: default
>   resourceVersion: "1467267"
>   uid: d2d51452-8a70-40d8-b61a-0a2e92f3039b

Reviewer: ```suggestion
# You can leave the existing metadata as they are.
# The values you'll see won't exactly match these.
metadata:
  creationTimestamp: "2023-12-27T08:24:51Z"
  name: sport
  namespace: default
  resourceVersion: "1467267"
  uid: d2d51452-8a70-40d8-b61a-0a2e92f3039b
```

Author: Thanks for the suggestion.

```

You should see the following output:
```
configmap/sport edited
```

Tail the logs of the Pod.
```shell

> kubectl logs -f <pod-name>

Reviewer: Does this work?
```suggestion
kubectl logs -f deployments/configmap-volume
```

Author: Yes, it works. I have implemented this in the latest commit.

```
After few seconds, you should see the log output change as follows:
```
FROM:
My preferred sport is football
TO:

> My preferred sport is cricket

Reviewer: You could try a `diff` format block. The `FROM:` and `TO:` aren't literal console output; don't make it look like that is what the reader would see.

Eg
```console
My preferred sport is football
My preferred sport is football
My preferred sport is football
My preferred sport is cricket
```

Extra credit: include a timestamp in the log lines.

Author: Done. It is implemented in the latest commit.

```

> When a ConfigMap mounted as a Volume is updated, the projected content is seamlessly updated too.

Reviewer: ```suggestion
When you have a ConfigMap that is mapped into a running Pod using either a `configMap` volume or a `projected` volume, and you update that ConfigMap, the running Pod sees the update almost immediately.
```

Author: Thanks for the suggestion.

kubectl get deployment configmap-env-var
```

You should see an output similar to:
```
NAME                READY   UP-TO-DATE   AVAILABLE   AGE
configmap-env-var   3/3     3            3           12m
```

Check the Pods.
```shell
kubectl get pods --selector=app=configmap-env-var
```

You should see an output similar to:
```
NAME                                 READY   STATUS        RESTARTS   AGE
configmap-env-var-6d94d89bf5-2ph2l   1/1     Running       0          13s
configmap-env-var-6d94d89bf5-74twx   1/1     Running       0          8s
configmap-env-var-6d94d89bf5-d5vx8   1/1     Running       0          11s
configmap-env-var-786b8bbfd6-hks7b   1/1     Terminating   0          100s
configmap-env-var-786b8bbfd6-jdzv6   1/1     Terminating   0          100s
configmap-env-var-786b8bbfd6-smlz5   1/1     Terminating   0          100s
```

View the logs of any running Pod.
```shell
kubectl logs <pod-name>
```

You should see the following output:
```
My preferred car is Tesla
```

This demonstrates the scenario of updating environment variables in a Pod that are derived from a ConfigMap. Changes to the ConfigMap values are applied to the Pod during the subsequent rollout.  

> **Summary:**  

Reviewer: ```suggestion
## Summary

```

---
title: Updating configuration via a ConfigMap

> content_type: tutorial

Reviewer: Great. Please also set a value for `weight` (should be different from all other weights within `content/en/docs/tutorials/configuration/`.

Author: Got it. I'll add a weight.

Author: Resolved.

---

<!-- overview -->

This page provides a step-by-step example updating configuration within a Pod via a ConfigMap and builds upon the [Configure a Pod to Use a ConfigMap](/docs/tasks/configure-pod-container/configure-pod-configmap/) task. 

> 

Reviewer: The example uses NGINX; I would mention NGINX here and also be clear that we chose that just as an example.

Author: Resolved.

## {{% heading "prerequisites" %}}
{{< include "task-tutorial-prereqs.md" >}}

## {{% heading "objectives" %}}
* Update configuration via a ConfigMap mounted as a Volume
* Update environment variables of a Pod via a ConfigMap 

<!-- lessoncontent -->

## Update configuration via a ConfigMap mounted as a Volume

Use the `kubectl create configmap` command to create a ConfigMap from [literal values](/docs/tasks/configure-pod-container/configure-pod-configmap/#create-configmaps-from-literal-values):

```shell
kubectl create configmap sport --from-literal=sport=football
```

> [Create a Deployment](/docs/concepts/workloads/controllers/deployment/#creating-a-deployment) manifest named `deployment.yaml` with the below specification. The ConfigMap `sport` is mounted as a Volume on the Pod.

Reviewer: Ideally, make this an example that people can apply using `kubectl apply -f <url>`. See other tutorials to see how this is done, or ask questions if you're still not sure.

Author: So I can add a new yaml to the `content/en/examples directory`, right?

Reviewer: You are welcome to do that.

Author: I guess I need some advice about this.
I added the yaml in the examples directory and it is rendered correctly.
`examples/deployments/deployment-with-configmap-as-volume.yaml`

Now as I understand, this yaml will be uploaded to the following location once the PR is merged to main, is that right?
https://raw.githubusercontent.com/kubernetes/website/main/content/en/examples/deployments/deployment-with-configmap-as-volume.yaml 

Currently this link throws a 404 as expected.

How do I validate that everything with this example yaml is correct before the PR is merged to main?

Author: This is resolved in the latest commit.

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: configmap-test
  labels:
    app: nginx
spec:

>   replicas: 1

Reviewer: BTW, it's more realistic to have at least 2 replicas.

Author: Acknowledged.

Author: Resolved, increased replicas to 3.

  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx

>         image: nginx:1.14.2

Reviewer: Consider using https://hub.docker.com/r/nginxinc/nginx-unprivileged, which does not require running as `root`.

Author: Clear, thanks for sending a reference to this image. I will use the nginx-unprivileged as an example.

Author: Resolved.

        ports:
        - containerPort: 80
        volumeMounts:
          - name: config-volume
            mountPath: /etc/config
      volumes:
        - name: config-volume
          configMap:

>             name: sport

Reviewer: How does this ConfigMap affect the behavior of the Pod?

Author: I will extend this to actually use the value in a log to stdout.

Author: I modified the tutorial to use the values from the ConfigMap in logs printed to stdout. This should be sufficient as I want to keep the application used in the tutorial as simple as possible.

```

Create the deployment.
```shell
kubectl apply -f deployment.yaml
```

Check the deployment.
```shell
kubectl get deployment configmap-test
```

You should see an output similar to:
```
NAME             READY   UP-TO-DATE   AVAILABLE   AGE
configmap-test   1/1     1            1           3s
```

Check the pod.
```shell

> kubectl get pod <pod-name>

Reviewer: How does the reader know what the name of the Pod is?

Author: Acknowledged. I will improve this.

Author: Resolved.

```

You should see the following output:
```
CAR=mercedes
```

Edit the ConfigMap.
```shell
kubectl edit configmap car
```

In the yaml editor, change the value of key `car` from `mercedes` to `tesla` and save.
```yaml
apiVersion: v1
data:
  car: tesla
kind: ConfigMap
metadata:
  creationTimestamp: "2023-12-27T09:02:47Z"
  name: car
  namespace: default
  resourceVersion: "1471221"
  uid: 3ce13cd2-dca1-4d3e-b157-af60b5797dab
```

You should see the following output:
```
configmap/car edited
```

View the value of the environment variable named `CAR`.
```shell
kubectl exec configmap-test-5fcb9c99d7-mgx9v -- env | grep CAR
```

You should see the following output:
```
CAR=mercedes
```

{{< note >}}
You will observe that although the value of the key inside the ConfigMap has changed, the environment variable in the Pod still possesses the earlier value. This is because environment variables in a Pod are not updated until the Pod is rolled-over.
{{< /note >}}

Let's simulate a roll-over for the Pod by adding an annotation to the Deployment's `template.metadata`.

Edit the deployment `configmap-test` to add an annotation to the Pod.
```shell

> kubectl edit deployment configmap-test

Reviewer: I'd use `kubectl rollout`.

Author: Agreed, this is a much better option. Thanks for pointing it out. I've resolved it.

---
title: Updating Configuration via a Configmap
content_type: tutorial

> weight: 20

Author: Ack, will be handled in the next commit.

---

<!-- overview -->
This page provides a step-by-step example of updating configuration within a Pod via a ConfigMap and builds upon the [Configure a Pod to Use a ConfigMap](/docs/tasks/configure-pod-container/configure-pod-configmap/) task. At the end of this tutorial, you will understand how to change the configuration for a running application. 
This tutorial uses the NGINX image as an example.

## {{% heading "prerequisites" %}}
{{< include "task-tutorial-prereqs.md" >}}

> 

Reviewer: Please mention that you need `curl` (it's really popular, but even so, let's be clear about what people need to have).

Author: Ack. I'm adding it as a new page under the `includes` directory to make it reusable. Any tutorial that needs `cURL` as a prerequisite can include this page in their tutorial to save time and have a standard pattern across tutorials.

## {{% heading "objectives" %}}
* Update configuration via a ConfigMap mounted as a Volume
* Update environment variables of a Pod via a ConfigMap 

<!-- lessoncontent -->

## Update configuration via a ConfigMap mounted as a Volume

Use the `kubectl create configmap` command to create a ConfigMap from [literal values](/docs/tasks/configure-pod-container/configure-pod-configmap/#create-configmaps-from-literal-values):

```shell
kubectl create configmap sport --from-literal=sport=football
```

> Below is an example of a Deployment manifest with the ConfigMap `sport` mounted as a Volume on the Pod.

Author: Ack. Thanks for this suggestion to use tooltips. I learnt something new. :)

{{% code_sample file="deployments/deployment-with-configmap-as-volume.yaml" %}}

> Create the deployment.

Reviewer: ```suggestion
Create the Deployment:
```

Author: Ack, will be included in the next commit.

```shell
kubectl apply -f https://k8s.io/examples/deployments/deployment-with-configmap-as-volume.yaml
```

> Check the deployment.

Reviewer: ```suggestion
Check the Deployment:
```

Author: Ack, will be included in the next commit.

```shell
kubectl get deployment configmap-volume
```

You should see an output similar to:
```
NAME               READY   UP-TO-DATE   AVAILABLE   AGE
configmap-volume   3/3     3            3           19s
```

> Check the pods.

Author: Ack, will be handled in the next commit.

```shell
kubectl get pods --selector=app=configmap-volume
```

You should see an output similar to:
```
NAME                                READY   STATUS    RESTARTS   AGE
configmap-volume-6b976dfdcf-qxvbm   1/1     Running   0          72s
configmap-volume-6b976dfdcf-skpvm   1/1     Running   0          72s
configmap-volume-6b976dfdcf-tbc6r   1/1     Running   0          72s
```

> The data in the ConfigMap is mounted to the Pods as a Volume. The Pod references this data to generate an output that is printed to stdout. Check this by viewing the logs of the Pods.

Reviewer: ```suggestion
On each node where one of these Pods is running, the kubelet fetches the data for that ConfigMap
and translates it to files in a local volume. The kubelet then mounts that volume into the container,
as specified in the Pod template. The code running in that container loads the information from the
file and uses it printed a report to stdout. You can check this report by viewing the logs for one of the
Pods in that Deployment:
```

Author: Ack, will be part of the next commit.

```shell

> kubectl logs deployments/configmap-volume

Reviewer: ```suggestion
# Pick one Pod that belongs to the Deployment, and view its logs
kubectl logs deployments/configmap-volume
```

Author: This should not be necessary as `kubectl logs deployments/configmap-volume` picks one of the Pod at random and displays its logs. This is seen in the output of the command, similar to `Found 3 pods, using pod/configmap-volume-76d9c5678f-x5rgj`

```

You should see an output similar to the below:
```
Found 3 pods, using pod/configmap-volume-76d9c5678f-x5rgj
Thu Jan  4 14:06:46 UTC 2024 My preferred sport is football
Thu Jan  4 14:06:56 UTC 2024 My preferred sport is football
Thu Jan  4 14:07:06 UTC 2024 My preferred sport is football
Thu Jan  4 14:07:16 UTC 2024 My preferred sport is football
Thu Jan  4 14:07:26 UTC 2024 My preferred sport is football
```

> Edit the ConfigMap.

Reviewer: ```suggestion
Edit the ConfigMap:
```

Author: Ack, will be handled in the next commit.

```shell
kubectl edit configmap sport
```

In the editor that appears, change the value of key `sport` from `football` to `cricket`. Save your changes.
The kubectl tool updates the ConfigMap accordingly (if you see an error, try again).

Here's an example of how that manifest could look after you edit it:

```yaml
apiVersion: v1
data:
  sport: cricket
kind: ConfigMap
# You can leave the existing metadata as they are.
# The values you'll see won't exactly match these.
metadata:
  creationTimestamp: "2024-01-04T14:05:06Z"
  name: sport
  namespace: default
  resourceVersion: "1743935"
  uid: 024ee001-fe72-487e-872e-34d6464a8a23
```

You should see the following output:
```
configmap/sport edited
```

> Tail the logs of the Deployment.

Reviewer: ```suggestion
Tail (follow the latest entries in) the logs of one of the pods that belongs to this Deployment:
```

Author: Ack, will be handled in the next commit.

```shell
kubectl logs -f deployments/configmap-volume
```
After few seconds, you should see the log output change as follows:
```
Thu Jan  4 14:11:36 UTC 2024 My preferred sport is football
Thu Jan  4 14:11:46 UTC 2024 My preferred sport is football
Thu Jan  4 14:11:56 UTC 2024 My preferred sport is football
Thu Jan  4 14:12:06 UTC 2024 My preferred sport is cricket
Thu Jan  4 14:12:16 UTC 2024 My preferred sport is cricket
```

> When you have a ConfigMap that is mapped into a running Pod using either a `configMap` volume or a `projected` volume, and you update that ConfigMap, the running Pod sees the update almost immediately.

Reviewer: ```suggestion
When you have a ConfigMap that is mapped into a running Pod using either a `configMap` volume
or a `projected` volume, and you update that ConfigMap, the running Pod sees the update almost
immediately. However, your application only sees the change if it is written to either poll for changes,
or watch for file updates. An application that loads its configuration once at startup will not notice
a change.
```

Author: Ack, very nice suggestion. It will be included in the next commit.

{{< note >}}
The total delay from the moment when the ConfigMap is updated to the moment when new keys are projected to the Pod can be as long as kubelet sync period.  
Also check [Mounted ConfigMaps are updated automatically](/docs/tasks/configure-pod-container/configure-pod-configmap/#mounted-configmaps-are-updated-automatically).
{{< /note >}}

## Update environment variables of a Pod via a ConfigMap
Use the `kubectl create configmap` command to create a ConfigMap from [literal values](/docs/tasks/configure-pod-container/configure-pod-configmap/#create-configmaps-from-literal-values):

```shell
kubectl create configmap fruits --from-literal=fruits=apples
```

Below is an example of a Deployment manifest with an environment variable configured via the ConfigMap `fruits`.
{{% code_sample file="deployments/deployment-with-configmap-as-envvar.yaml" %}}

> Create the deployment.

Reviewer: (nit)
```suggestion
Create the Deployment:
```

Author: Ack, will be part of the next commit.

```shell
kubectl apply -f https://k8s.io/examples/deployments/deployment-with-configmap-as-envvar.yaml
```

> Check the deployment.

Reviewer: ```suggestion
Check the Deployment:
```

Author: Ack, will be part of the next commit.

```shell
kubectl get deployment configmap-env-var
```

You should see an output similar to:
```
NAME                READY   UP-TO-DATE   AVAILABLE   AGE
configmap-env-var   3/3     3            3           7s
```

> Check the pods.

Reviewer: ```suggestion
Check the pods for this Deployment (matching by
{{< glossary_tooltip text="selector" term_id="selector" >}}):
```

Author: Ack, will be part of the next commit.

```shell
kubectl get pods --selector=app=configmap-env-var
```

You should see an output similar to:
```
NAME                                 READY   STATUS    RESTARTS   AGE
configmap-env-var-59cfc64f7d-74d7z   1/1     Running   0          46s
configmap-env-var-59cfc64f7d-c4wmj   1/1     Running   0          46s
configmap-env-var-59cfc64f7d-dpr98   1/1     Running   0          46s
```

The key-value pair in the ConfigMap is configured as an environment variable in the container of the Pod.
Check this by viewing the logs of the Pod.
```shell
kubectl logs deployment/configmap-env-var
```

> You should see an output similar to the below:

Reviewer: More idiomatic as English:
```suggestion
You should see an output similar to:
```

Author: Ack, will be part of the next commit.

```
Found 3 pods, using pod/configmap-env-var-7c994f7769-l74nq
Thu Jan  4 16:07:06 UTC 2024 The basket is full of apples
Thu Jan  4 16:07:16 UTC 2024 The basket is full of apples
Thu Jan  4 16:07:26 UTC 2024 The basket is full of apples
```

> Edit the ConfigMap.

Reviewer: ```suggestion
Edit the ConfigMap:
```

Author: Ack, will be part of the next commit.

```shell
kubectl edit configmap fruits
```

In the editor that appears, change the value of key `fruits` from `apples` to `mangoes`. Save your changes.
The kubectl tool updates the ConfigMap accordingly (if you see an error, try again).

Here's an example of how that manifest could look after you edit it:
```yaml
apiVersion: v1
data:
  fruits: mangoes
kind: ConfigMap
# You can leave the existing metadata as they are.
# The values you'll see won't exactly match these.
metadata:
  creationTimestamp: "2024-01-04T16:04:19Z"
  name: fruits
  namespace: default
  resourceVersion: "1749472"

```

You should see the following output:
```
configmap/fruits edited
```

Tail the logs of the Deployment and observe the output for few seconds:
```shell
kubectl logs deployment/configmap-env-var
```

> You will notice that the output remains unchanged:

Reviewer: ```suggestion
Notice that the output remains **unchanged**, even though you edited the ConfigMap:
```

Author: Ack, will be part of the next commit.

```
Thu Jan  4 16:12:56 UTC 2024 The basket is full of apples
Thu Jan  4 16:13:06 UTC 2024 The basket is full of apples
Thu Jan  4 16:13:16 UTC 2024 The basket is full of apples
Thu Jan  4 16:13:26 UTC 2024 The basket is full of apples

```

{{< note >}}

> Although the value of the key inside the ConfigMap has changed, the environment variable in the Pod still possesses the earlier value. This is because environment variables in a Pod are not updated until the Pod is rolled-over.

Reviewer: ```suggestion
Although the value of the key inside the ConfigMap has changed, the environment variable in the Pod still
shows the earlier value. This is because environment variables for a process running inside a Pod are
**not** updated when the source data changes; if you wanted to force an update, you would need to
have Kubernetes replace your existing Pods. The new Pods would then run with the updated information.
```

Author: Ack, will be included in the next commit.

{{< /note >}}

> Let's perform a rollout for the Deployment that will cause all the Pods to restart.

Reviewer: ```suggestion
You can trigger that replacement. Perform a rollout for the Deployment, using
[`kubectl rollout`](/docs/reference/kubectl/generated/kubectl_rollout/):
```

Author: Ack, will be included in the next commit.

```shell

> kubectl rollout restart deployment configmap-env-var

Reviewer: ```suggestion
# Trigger the rollout
kubectl rollout restart deployment configmap-env-var

# Wait for the rollout to complete
kubectl rollout status deployment configmap-env-var --watch=true
```

Author: Ack, will be part of the next commit.

```

> Check the deployment.

Reviewer: ```suggestion
Next, check the Deployment:
```

Author: Ack, will be part of the next commit.

```shell
kubectl get deployment configmap-env-var
```

You should see an output similar to:
```
NAME                READY   UP-TO-DATE   AVAILABLE   AGE
configmap-env-var   3/3     3            3           12m
```

> Check the Pods.

Reviewer: ```suggestion
Check the Pods:
```

Author: Ack, will be part of the next commit.

```shell
kubectl get pods --selector=app=configmap-env-var
```

> The older Pods will terminate and new ones will be created. You should finally see an output similar to:

Reviewer: ```suggestion
The rollout caused Kubernetes to make a new
{{< glossary_tooltip term_id="replica-set" text="ReplicaSet" >}} for the Deployment; that means that the
existing Pods were terminated, and new ones have be created. You should finally see an output similar to:
```

Author: Ack, will be part of the next commit.

```
NAME                                 READY   STATUS        RESTARTS   AGE
configmap-env-var-6d94d89bf5-2ph2l   1/1     Running       0          13s
configmap-env-var-6d94d89bf5-74twx   1/1     Running       0          8s
configmap-env-var-6d94d89bf5-d5vx8   1/1     Running       0          11s
```

> View the logs of the Deployment.

Reviewer: ```suggestion
View the logs for a Pod in this Deployment:
```

Author: Ack

```shell

> kubectl logs deployment/configmap-env-var

Reviewer: ```suggestion
# Pick one Pod that belongs to the Deployment, and view its logs
kubectl logs deployment/configmap-env-var
```

Author: This should not be necessary as kubectl logs deployments/configmap-volume picks one of the Pod at random and displays its logs.

Reviewer: We're explaining what the command does, here.

```

You should see an output similar to the below:
```
Found 3 pods, using pod/configmap-env-var-6d9ff89fb6-bzcf6
Thu Jan  4 16:30:35 UTC 2024 The basket is full of mangoes
Thu Jan  4 16:30:45 UTC 2024 The basket is full of mangoes
Thu Jan  4 16:30:55 UTC 2024 The basket is full of mangoes
```

> This demonstrates the scenario of updating environment variables in a Pod that are derived from a ConfigMap. Changes to the ConfigMap values are applied to the Pod during the subsequent rollout.  

Reviewer: ```suggestion
This demonstrates the scenario of updating environment variables in a Pod that are derived
from a ConfigMap. Changes to the ConfigMap values are applied to the Pod during the subsequent
rollout. If Pods get created for other reason, such as scaling up the Deployment, then the new Pods
also use the latest configuration values; if you don't trigger a rollout, then you might find that your
app is running with a mix of old and new environment variable values.
```

Author: Great addition. Thanks. Ack.

## Update configuration via a ConfigMap in a multi-container Pod

Use the `kubectl create configmap` command to create a ConfigMap from [literal values](/docs/tasks/configure-pod-container/configure-pod-configmap/#create-configmaps-from-literal-values):

```shell

> kubectl create configmap colour --from-literal=colour=red

Author: Ack, will be handled in the next commit.

```

> Below is an example of a Deployment having Pods with two containers. The two containers share an ```emptyDir``` Volume that they use to communicate. The first container runs a ```nginx``` server. The mount path for the shared Volume is ```/usr/share/nginx/html```. The second container is based on the ```alpine``` image, and has a mount path of ```/pod-data```. The ```alpine``` container writes a file in HTML which has its content based on a ConfigMap. The ```nginx``` container renders the HTML via HTTP.

Author: Ack, will be handled in the next commit.

{{% code_sample file="deployments/deployment-with-configmap-two-containers.yaml" %}}

> Create the deployment.

Reviewer: ```suggestion
Create the Deployment:
```

Author: Ack, will be handled in the next commit.

```shell
kubectl apply -f https://k8s.io/examples/deployments/deployment-with-configmap-two-containers.yaml
```

> Check the deployment.

Reviewer: ```suggestion
Check the Deployment:
```

Author: Ack, will be handled in the next commit.

```shell
kubectl get deployment configmap-two-containers
```

You should see an output similar to:
```shell
NAME                       READY   UP-TO-DATE   AVAILABLE   AGE
configmap-two-containers   3/3     3            3           12s

```

> Check the pods.

Reviewer: ```suggestion
Check the pods:
```

Author: Ack, will be handled in the next commit.

```shell
kubectl get pods --selector=app=configmap-two-containers
```
You should see an output similar to:

> ```shell

Reviewer: This output isn't a shell script; drop the `shell` part.

Author: Ack, will be handled in the next commit.

NAME                                        READY   STATUS    RESTARTS   AGE
configmap-two-containers-565fb6d4f4-2xhxf   2/2     Running   0          20s
configmap-two-containers-565fb6d4f4-g5v4j   2/2     Running   0          20s
configmap-two-containers-565fb6d4f4-mzsmf   2/2     Running   0          20s
````

> Expose the deployment via a Service.

Reviewer: ```suggestion
Expose the Deployment (the `kubectl` tool creates a
{{<glossary_tooltip text="Service" term_id="service">}}  for you):
```

Author: Ack, will be handled in the next commit.

```shell
kubectl expose deployment configmap-two-containers --name=configmap-service --port=8080 --target-port=80
```

Use ```kubectl``` to forward the port:
```shell

> kubectl port-forward service/configmap-service 8080:8080 &

Reviewer: ```suggestion
kubectl port-forward service/configmap-service 8080:8080 & # this stays running in the background
```

```

Access the service.
```shell

> curl localhost:8080

Reviewer: ```suggestion
curl http://localhost:8080
```

Author: Ack, will be handled in the next commit.

```

You should see an output similar to:

> ```shell

Reviewer: This output isn't a shell script; drop the `shell` part.

Author: Ack, will be handled in the next commit.

Fri Jan  5 08:08:22 UTC 2024 My preferred colour is red
```

> Edit the ConfigMap.

Author: Ack, will be handled in the next commit.

```shell
kubectl edit configmap colour
```

In the editor that appears, change the value of key `colour` from `red` to `blue`. Save your changes.
The kubectl tool updates the ConfigMap accordingly (if you see an error, try again).

Here's an example of how that manifest could look after you edit it:

```yaml
apiVersion: v1
data:
  colour: blue
kind: ConfigMap
# You can leave the existing metadata as they are.
# The values you'll see won't exactly match these.
metadata:
  creationTimestamp: "2024-01-05T08:12:05Z"
  name: colour
  namespace: configmap
  resourceVersion: "1801272"
  uid: 80d33e4a-cbb4-4bc9-ba8c-544c68e425d6
```

Loop over the service URL for few seconds. 
```shell

> while true; do curl localhost:8080; sleep 10; done

Reviewer: How about:
```suggestion
# Cancel this when you're happy with it (Ctrl-C)
while true; do curl --connect-timeout 7.5 http://localhost:8080; sleep 10; done
```

Author: Ack, will be handled in the next commit.

```

You should see the output change as follows:

> ```shell

Author: Ack, will be handled in the next commit.

Fri Jan  5 08:14:00 UTC 2024 My preferred colour is red
Fri Jan  5 08:14:02 UTC 2024 My preferred colour is red
Fri Jan  5 08:14:20 UTC 2024 My preferred colour is red
Fri Jan  5 08:14:22 UTC 2024 My preferred colour is red
Fri Jan  5 08:14:32 UTC 2024 My preferred colour is blue
Fri Jan  5 08:14:43 UTC 2024 My preferred colour is blue
Fri Jan  5 08:15:00 UTC 2024 My preferred colour is blue
```

> 

Author: I will work on adding this, however it may need some time. Once its ready, we can add it to the current PR if the PR is still open. Else, I can create a new PR for it later.

I believe this should not be a blocker to merge the current PR.

> 

Author: Added a 4th scenario with the Sidecar Container.

---
reviewers:
  - sftim
title: Updating Configuration via a Configmap
content_type: tutorial
weight: 20
---

<!-- overview -->
This page provides a step-by-step example of updating configuration within a Pod via a ConfigMap and builds upon the [Configure a Pod to Use a ConfigMap](/docs/tasks/configure-pod-container/configure-pod-configmap/) task. At the end of this tutorial, you will understand how to change the configuration for a running application. 
This tutorial uses the NGINX image as an example.

## {{% heading "prerequisites" %}}
{{< include "task-tutorial-prereqs.md" >}}

> {{< include "task-tutorial-curl-prereq.md" >}}

Author: Thanks, I have implemented this.

Found 3 pods, using pod/configmap-volume-76d9c5678f-x5rgj
Thu Jan  4 14:06:46 UTC 2024 My preferred sport is football
Thu Jan  4 14:06:56 UTC 2024 My preferred sport is football
Thu Jan  4 14:07:06 UTC 2024 My preferred sport is football
Thu Jan  4 14:07:16 UTC 2024 My preferred sport is football
Thu Jan  4 14:07:26 UTC 2024 My preferred sport is football
```

Edit the ConfigMap:
```shell
kubectl edit configmap sport
```

In the editor that appears, change the value of key `sport` from `football` to `cricket`. Save your changes.
The kubectl tool updates the ConfigMap accordingly (if you see an error, try again).

Here's an example of how that manifest could look after you edit it:

```yaml
apiVersion: v1
data:
  sport: cricket
kind: ConfigMap
# You can leave the existing metadata as they are.
# The values you'll see won't exactly match these.
metadata:
  creationTimestamp: "2024-01-04T14:05:06Z"
  name: sport
  namespace: default
  resourceVersion: "1743935"
  uid: 024ee001-fe72-487e-872e-34d6464a8a23
```

You should see the following output:
```
configmap/sport edited
```

Tail (follow the latest entries in) the logs of one of the pods that belongs to this Deployment:
```shell

> kubectl logs -f deployments/configmap-volume

Reviewer: I recommend
```suggestion
kubectl logs deployments/configmap-volume --follow
```

Check this by viewing the logs of the Pod.
```shell
kubectl logs deployment/configmap-env-var
```

You should see an output similar to:
```
Found 3 pods, using pod/configmap-env-var-7c994f7769-l74nq
Thu Jan  4 16:07:06 UTC 2024 The basket is full of apples
Thu Jan  4 16:07:16 UTC 2024 The basket is full of apples
Thu Jan  4 16:07:26 UTC 2024 The basket is full of apples
```

Edit the ConfigMap:
```shell
kubectl edit configmap fruits
```

In the editor that appears, change the value of key `fruits` from `apples` to `mangoes`. Save your changes.
The kubectl tool updates the ConfigMap accordingly (if you see an error, try again).

Here's an example of how that manifest could look after you edit it:
```yaml
apiVersion: v1
data:
  fruits: mangoes
kind: ConfigMap
# You can leave the existing metadata as they are.
# The values you'll see won't exactly match these.
metadata:
  creationTimestamp: "2024-01-04T16:04:19Z"
  name: fruits
  namespace: default
  resourceVersion: "1749472"

```

You should see the following output:
```
configmap/fruits edited
```

Tail the logs of the Deployment and observe the output for few seconds:
```shell

> kubectl logs deployment/configmap-env-var

Reviewer: I recommend
```suggestion
# As the text explains, the output does NOT change
kubectl logs deployments/configmap-env-var --follow
```

```

Notice that the output remains **unchanged**, even though you edited the ConfigMap:
```
Thu Jan  4 16:12:56 UTC 2024 The basket is full of apples
Thu Jan  4 16:13:06 UTC 2024 The basket is full of apples
Thu Jan  4 16:13:16 UTC 2024 The basket is full of apples
Thu Jan  4 16:13:26 UTC 2024 The basket is full of apples

```

{{< note >}}
Although the value of the key inside the ConfigMap has changed, the environment variable in the Pod still shows the earlier value. This is because environment variables for a process running inside a Pod are **not** updated when the source data changes; if you wanted to force an update, you would need to have Kubernetes replace your existing Pods. The new Pods would then run with the updated information.
{{< /note >}}

You can trigger that replacement. Perform a rollout for the Deployment, using
[`kubectl rollout`](/docs/reference/kubectl/generated/kubectl_rollout/):
```shell
# Trigger the rollout
kubectl rollout restart deployment configmap-env-var

# Wait for the rollout to complete
kubectl rollout status deployment configmap-env-var --watch=true
```

> 

Reviewer: There's a problem; `kubectl rollout` does not actually wait for the old Pods to disappear. The old Pods will still be around, for a short time, albeit `Terminating`.

Reviewer: This works, but what a lot to type:
```shell
kubectl rollout restart deployment configmap-env-var && \
  kubectl rollout status deployment configmap-env-var --watch=true && \
  while kubectl get pods --selector=app.kubernetes.io/name=configmap-env-var \
  --template '{{range .items}}{{ if .metadata.deletionTimestamp }}pods/{{.metadata.name}}{{"\n"}}{{end}}{{end}}' |
  grep -q . ; do sleep 1.5; done && echo Rollout complete 1>&2
```

(I wonder if `kubectl rollout` should support waiting for the selected Pods to actually terminate, or if the Deployment and ReplicaSet controllers should somehow track the Pods that they once owned right through to deletion)

Author: Thanks for the alternate command, but I agree it is too much to type for this tutorial and we would not want to confuse the readers.

I will handle it by modifying the text to clearly imply that the readers should wait till the older Pods terminate completely before proceeding with the tutorial.

Next, check the Deployment:
```shell
kubectl get deployment configmap-env-var
```

You should see an output similar to:
```
NAME                READY   UP-TO-DATE   AVAILABLE   AGE
configmap-env-var   3/3     3            3           12m
```

Check the Pods:
```shell
kubectl get pods --selector=app.kubernetes.io/name=configmap-env-var
```

The rollout caused Kubernetes to make a new {{< glossary_tooltip term_id="replica-set" text="ReplicaSet" >}} for the Deployment; that means that the existing Pods were terminated, and new ones have been created. You should finally see an output similar to:
```
NAME                                 READY   STATUS        RESTARTS   AGE
configmap-env-var-6d94d89bf5-2ph2l   1/1     Running       0          13s
configmap-env-var-6d94d89bf5-74twx   1/1     Running       0          8s
configmap-env-var-6d94d89bf5-d5vx8   1/1     Running       0          11s
```

> View the logs for a Pod in this Deployment:

Reviewer: There's a problem: we haven't waited for the previous Pods to terminate.

This works, but it's a bit long to type:
```shell
kubectl logs --follow \
  "$( kubectl get pods --selector=app.kubernetes.io/name=configmap-env-var \
  --template '{{range .items}}{{ if not .metadata.deletionTimestamp }}pods/{{.metadata.name}}{{"\n"}}{{end}}{{end}}' |
  head -n 1 )"
```

Author: Thanks for the alternate command, but I agree it is too much to type for this tutorial and we would not want to confuse the readers.

I will handle it by modifying the text to clearly imply that the readers should wait till the older Pods terminate completely before proceeding with the tutorial.

Reviewer: I've marked this unresolved until a fix is confirmed.

Author: I've addressed this in the latest commit.

immutable-configmap-volume-78b6fbff95-vdslm   1/1     Running   0          62s
```

The Pod's container refers to the data defined in the ConfigMap and uses it to print a report to stdout. You can check this report by viewing the logs for one of the Pods in that Deployment:
```shell
# Pick one Pod that belongs to the Deployment, and view its logs
kubectl logs deployments/immutable-configmap-volume
```

You should see an output similar to:
```
Found 3 pods, using pod/immutable-configmap-volume-78b6fbff95-5gsfh
Wed Mar 20 03:52:34 UTC 2024 The name of the company is ACME, Inc.
Wed Mar 20 03:52:44 UTC 2024 The name of the company is ACME, Inc.
Wed Mar 20 03:52:54 UTC 2024 The name of the company is ACME, Inc.
```

{{< note >}}
Once a ConfigMap is marked as immutable, it is not possible to revert this change nor to mutate the contents of the data or the binaryData field.  
In order to modify the contents, we will change the name of the ConfigMap, and switch to running Pods that reference the new name.
{{< /note >}}

Edit the ConfigMap:
```shell
kubectl edit configmap company-name-20150801
```

In the editor that appears, perform the following actions:

> - change the name of the ConfigMap from `company-name-20150801` to `company-name-20240312`

Reviewer: does this really work?

Author: All instructions have been tested by me and they work successfully.

Use the `kubectl create configmap` command to create a ConfigMap from [literal values](/docs/tasks/configure-pod-container/configure-pod-configmap/#create-configmaps-from-literal-values):

```shell
kubectl create configmap sport --from-literal=sport=football
```
Below is an example of a Deployment manifest with the ConfigMap `sport` mounted as a
{{< glossary_tooltip text="volume" term_id="volume" >}} into the Pod's only container.
{{% code_sample file="deployments/deployment-with-configmap-as-volume.yaml" %}}

Create the Deployment:
```shell
kubectl apply -f https://k8s.io/examples/deployments/deployment-with-configmap-as-volume.yaml
```

Check the pods for this Deployment to ensure they are ready (matching by
{{< glossary_tooltip text="selector" term_id="selector" >}}):
```shell
kubectl get pods --selector=app.kubernetes.io/name=configmap-volume
```

You should see an output similar to:
```
NAME                                READY   STATUS    RESTARTS   AGE
configmap-volume-6b976dfdcf-qxvbm   1/1     Running   0          72s
configmap-volume-6b976dfdcf-skpvm   1/1     Running   0          72s
configmap-volume-6b976dfdcf-tbc6r   1/1     Running   0          72s
```

> On each node where one of these Pods is running, the kubelet fetches the data for that ConfigMap and translates it to files in a local volume. The kubelet then mounts that volume into the container, as specified in the Pod template. The code running in that container loads the information from the file and uses it to print a report to stdout. You can check this report by viewing the logs for one of the Pods in that Deployment:

Reviewer 3: Please do wrap long lines like this.

Author: Ack. This will be handled in the next commit.

  creationTimestamp: "2024-01-04T14:05:06Z"
  name: sport
  namespace: default
  resourceVersion: "1743935"
  uid: 024ee001-fe72-487e-872e-34d6464a8a23
```

You should see the following output:
```
configmap/sport edited
```

Tail (follow the latest entries in) the logs of one of the pods that belongs to this Deployment:
```shell
kubectl logs deployments/configmap-volume --follow
```
After few seconds, you should see the log output change as follows:
```
Thu Jan  4 14:11:36 UTC 2024 My preferred sport is football
Thu Jan  4 14:11:46 UTC 2024 My preferred sport is football
Thu Jan  4 14:11:56 UTC 2024 My preferred sport is football
Thu Jan  4 14:12:06 UTC 2024 My preferred sport is cricket
Thu Jan  4 14:12:16 UTC 2024 My preferred sport is cricket
```

When you have a ConfigMap that is mapped into a running Pod using either a 
`configMap` volume or a `projected` volume, and you update that ConfigMap, 
the running Pod sees the update almost immediately.  
However, your application only sees the change if it is written to either poll for changes, 
or watch for file updates.  
An application that loads its configuration once at startup will not notice a change.

{{< note >}}

> The total delay from the moment when the ConfigMap is updated to the moment when new keys are projected to the Pod can be as long as kubelet sync period.  

Reviewer 3: Please wrap the long lines.

Author: Ack. This will be handled in the next commit.

Also check [Mounted ConfigMaps are updated automatically](/docs/tasks/configure-pod-container/configure-pod-configmap/#mounted-configmaps-are-updated-automatically).
{{< /note >}}

> ## Update environment variables of a Pod via a ConfigMap {#rollout-configmap-env}

Reviewer 3: I don't think we need this subsection. In a tutorial, we show users how to get something done, step by step. It is weird to read something like "hey, follow me, I'm showing you that something doesn't work."

It is for sure that we need a callout, maybe something like a warning, to tell users that this update cannot apply to environment variables. We can briefly explain why it doesn't work. However, having a dedicated subsection here showing that something doesn't work sounds an overkill.

Author: Once a reader follows these steps, the difference between ConfigMap as Volume and as env variable will be crystal clear. I think the section has merit, and reducing it to a callout may be insufficient to demonstrate the difference.

I propose to keep the section, in any case it will not do any harm :)

{{% code_sample file="deployments/deployment-with-configmap-two-containers.yaml" %}}

Create the Deployment:
```shell
kubectl apply -f https://k8s.io/examples/deployments/deployment-with-configmap-two-containers.yaml
```

Check the pods for this Deployment to ensure they are ready (matching by
{{< glossary_tooltip text="selector" term_id="selector" >}}):
```shell
kubectl get pods --selector=app.kubernetes.io/name=configmap-two-containers
```
You should see an output similar to:
```
NAME                                        READY   STATUS    RESTARTS   AGE
configmap-two-containers-565fb6d4f4-2xhxf   2/2     Running   0          20s
configmap-two-containers-565fb6d4f4-g5v4j   2/2     Running   0          20s
configmap-two-containers-565fb6d4f4-mzsmf   2/2     Running   0          20s
````

Expose the Deployment (the `kubectl` tool creates a
{{<glossary_tooltip text="Service" term_id="service">}}  for you):
```shell
kubectl expose deployment configmap-two-containers --name=configmap-service --port=8080 --target-port=80
```

> Use ```kubectl``` to forward the port:

Reviewer: (nit)
```suggestion
Use `kubectl` to forward the port:
```

Author: Ack. Will be handled in the next commit.

```shell
kubectl port-forward service/configmap-service 8080:8080 & # this stays running in the background
```

Access the service.
```shell
curl http://localhost:8080
```

> 

Reviewer 3: Is this Service creation relevant to the topic?

Reviewer: It lets people see the impact of the ConfigMap change, right?

Author: Yes, the service creation is relevant. The section on multi-container pods is serving an html page on nginx. To enable http requests from outside the cluster we're creating a ClusterIP service and using the kubectl port-forward functionality to reach it. On minikube, I prefer using the port-forwarding than creating a NodePort service.

You should see an output similar to:
```
Fri Jan  5 08:08:22 UTC 2024 My preferred color is red
```

Edit the ConfigMap:
```shell
kubectl edit configmap color
```

In the editor that appears, change the value of key `color` from `red` to `blue`. Save your changes.
The kubectl tool updates the ConfigMap accordingly (if you see an error, try again).

Here's an example of how that manifest could look after you edit it:

```yaml
apiVersion: v1
data:
  color: blue
kind: ConfigMap
# You can leave the existing metadata as they are.
# The values you'll see won't exactly match these.
metadata:
  creationTimestamp: "2024-01-05T08:12:05Z"
  name: color
  namespace: configmap
  resourceVersion: "1801272"
  uid: 80d33e4a-cbb4-4bc9-ba8c-544c68e425d6
```

Loop over the service URL for few seconds. 
```shell
# Cancel this when you're happy with it (Ctrl-C)

> while true; do curl --connect-timeout 7.5 http://localhost:8080; sleep 10; done

Reviewer: Where do you run this? Does the `localhost` name resolve where people would expect?

Author: This is run from a local terminal (through which all the kubectl commands are executed). The `localhost` name resolves perfectly. I've tested it via minikube.

```

You should see the output change as follows:
```
Fri Jan  5 08:14:00 UTC 2024 My preferred color is red
Fri Jan  5 08:14:02 UTC 2024 My preferred color is red
Fri Jan  5 08:14:20 UTC 2024 My preferred color is red
Fri Jan  5 08:14:22 UTC 2024 My preferred color is red
Fri Jan  5 08:14:32 UTC 2024 My preferred color is blue
Fri Jan  5 08:14:43 UTC 2024 My preferred color is blue
Fri Jan  5 08:15:00 UTC 2024 My preferred color is blue
```

## Update configuration via a ConfigMap in a Pod possessing a sidecar container {#rollout-configmap-sidecar}

The above scenario can be replicated by using a [Sidecar Container](/docs/concepts/workloads/pods/sidecar-containers/) as a helper container to write the HTML file.  

> As a Sidecar Container is conceptually an Init Container, it is guaranteed to start before the main webserver container.  

Reviewer 3: webserver is not an English word.

Reviewer: (IMO, we can merge this PR even with the “webserver” word as is. I'm a native English speaker and I write “webserver” without the space fairly often).

Author: I will make the change from "webserver" to "web server" in the next commit.

This ensures that the HTML file is always available when the webserver is ready to serve it.  
Please see [Enabling sidecar containers](/docs/concepts/workloads/pods/sidecar-containers/#enabling-sidecar-containers) to utilize this feature.

If you are continuing from the previous scenario, you can reuse the ConfigMap named `color` for this scenario.  
If you are executing this scenario independently, use the `kubectl create configmap` command to create a ConfigMap from [literal values](/docs/tasks/configure-pod-container/configure-pod-configmap/#create-configmaps-from-literal-values):

```shell
kubectl create configmap color --from-literal=color=blue
```

> Below is an example manifest for a Deployment that manages a set of Pods, each with a main container and a Sidecar Container. The two containers share an `emptyDir` volume that they use to communicate.

Reviewer 3: Sidecar should be spelt as "sidecar" ?

Author: Ack. This will be handled in the next commit.

Create the Deployment:
```shell
kubectl apply -f https://k8s.io/examples/deployments/deployment-with-configmap-and-sidecar-container.yaml
```

Check the pods for this Deployment to ensure they are ready (matching by
{{< glossary_tooltip text="selector" term_id="selector" >}}):
```shell
kubectl get pods --selector=app.kubernetes.io/name=configmap-sidecar-container
```
You should see an output similar to:
```
NAME                                           READY   STATUS    RESTARTS   AGE
configmap-sidecar-container-5fb59f558b-87rp7   2/2     Running   0          94s
configmap-sidecar-container-5fb59f558b-ccs7s   2/2     Running   0          94s
configmap-sidecar-container-5fb59f558b-wnmgk   2/2     Running   0          94s
````

Expose the Deployment (the `kubectl` tool creates a
{{<glossary_tooltip text="Service" term_id="service">}}  for you):
```shell
kubectl expose deployment configmap-sidecar-container --name=configmap-sidecar-service --port=8081 --target-port=80
```

Use ```kubectl``` to forward the port:
```shell
kubectl port-forward service/configmap-sidecar-service 8081:8081 & # this stays running in the background
```

Access the service.
```shell

> curl http://localhost:8081

Reviewer: Where do you run this? Does the `localhost` name resolve where people would expect?

Author: This is run from a local terminal (through which all the kubectl commands are executed). The localhost name resolves perfectly. I've tested it via minikube.

```

You should see an output similar to:
```
Sat Feb 17 13:09:05 UTC 2024 My preferred color is blue
```

Edit the ConfigMap:
```shell
kubectl edit configmap color
```

In the editor that appears, change the value of key `color` from `blue` to `green`. Save your changes.
The kubectl tool updates the ConfigMap accordingly (if you see an error, try again).

Here's an example of how that manifest could look after you edit it:

```yaml
apiVersion: v1
data:
  color: green
kind: ConfigMap
# You can leave the existing metadata as they are.
# The values you'll see won't exactly match these.
metadata:
  creationTimestamp: "2024-02-17T12:20:30Z"
  name: color
  namespace: default
  resourceVersion: "1054"
  uid: e40bb34c-58df-4280-8bea-6ed16edccfaa
```

Loop over the service URL for few seconds.
```shell
# Cancel this when you're happy with it (Ctrl-C)

> while true; do curl --connect-timeout 7.5 http://localhost:8081; sleep 10; done

Reviewer: Where do you run this? Does the `localhost` name resolve where people would expect?

Author: This is run from a local terminal (through which all the kubectl commands are executed). The localhost name resolves perfectly. I've tested it via minikube.

```

You should see the output change as follows:
```
Sat Feb 17 13:12:35 UTC 2024 My preferred color is blue
Sat Feb 17 13:12:45 UTC 2024 My preferred color is blue
Sat Feb 17 13:12:55 UTC 2024 My preferred color is blue
Sat Feb 17 13:13:05 UTC 2024 My preferred color is blue
Sat Feb 17 13:13:15 UTC 2024 My preferred color is green
Sat Feb 17 13:13:25 UTC 2024 My preferred color is green
Sat Feb 17 13:13:35 UTC 2024 My preferred color is green
```

> ## Update configuration via an immutable ConfigMap that is mounted as a Volume {#rollout-configmap-immutable-volume}

Reviewer: nit: I wouldn't capitalize Volume as it is not an API kind.

Author: Ack. This will be handled in the next commit.

```yaml
apiVersion: v1
data:
  company_name: "Fiktivesunternehmen GmbH" # new fictional company name
immutable: true
kind: ConfigMap
# Apart from the name field, you can leave the existing metadata as they are.
# The values you'll see won't exactly match these.
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","data":{"company_name":"ACME, Inc."},"immutable":true,"kind":"ConfigMap","metadata":{"annotations":{},"name":"company-name-20150801","namespace":"default"}}
  creationTimestamp: "2024-03-20T03:51:47Z"
  name: company-name-20240312 # new ConfigMap name
  namespace: default
  resourceVersion: "854"
  uid: bc6d7c83-1b01-43d2-acb9-c4e1a4e4a8e8
```

Attempt to save the changes. This action is refused as the name of the ConfigMap was changed. Instead, a copy of the changes is stored in a temporary manifest as shown below:

```
A copy of your changes has been stored to "/var/folders/nn/bb6y1j8d69n9lhqppjlrfthm0000gn/T/kubectl-edit-1440698061.yaml"
error: At least one of apiVersion, kind and name was changed
```

Create a new ConfigMap by referencing the temporary manifest.
```shell

> kubectl apply -f /var/folders/nn/bb6y1j8d69n9lhqppjlrfthm0000gn/T/kubectl-edit-1440698061.yaml

Reviewer 3: Engineering practice wise, I don't think we would encourage users to create a resource out of a temporary file. The result might be correct, but the practice should be discouraged.

Reviewer: I agree; it's better to either:
- teach people to fetch the ConfigMap as YAML and edit it and apply the new manifest
- explain making a new immutable ConfigMap from a new manifest

I personally would also delete the old (immutable) ConfigMap once there's evidence that nothing is using it.

Author: Thanks for the advise. I have fixed this in the latest commit.

```

You should see an output similar to:
```
configmap/company-name-20240312 created

> ```

Reviewer: Should they delete the old ConfigMap?

Author: It is not necessary to delete the old ConfigMap.

---
reviewers:
  - sftim
title: Updating Configuration via a Configmap
content_type: tutorial
weight: 20
---

<!-- overview -->
This page provides a step-by-step example of updating configuration within a Pod via a ConfigMap and builds upon the [Configure a Pod to Use a ConfigMap](/docs/tasks/configure-pod-container/configure-pod-configmap/) task. At the end of this tutorial, you will understand how to change the configuration for a running application. 
This tutorial uses the alpine and NGINX images as examples.

## {{% heading "prerequisites" %}}
{{< include "task-tutorial-prereqs.md" >}}

You need to have the [Curl](https://curl.se/) command-line tool for making HTTP requests from
the terminal or command prompt. If you do not have `curl` available, you can install it. Check the
documentation for your local operating system.

## {{% heading "objectives" %}}
* Update configuration via a ConfigMap mounted as a Volume
* Update environment variables of a Pod via a ConfigMap
* Update configuration via a ConfigMap in a multi-container Pod
* Update configuration via a ConfigMap in a Pod possessing a Sidecar Container

<!-- lessoncontent -->

> ## Update configuration via a ConfigMap mounted as a Volume

Reviewer: ```suggestion
## Update configuration via a ConfigMap mounted as a Volume {#rollout-configmap-volume}
```

configmap-env-var-6d94d89bf5-74twx   1/1     Running       0          8s
configmap-env-var-6d94d89bf5-d5vx8   1/1     Running       0          11s
```

{{< note >}}
Please wait for the older Pods to fully terminate before proceeding with the next steps.
{{< /note >}}

View the logs for a Pod in this Deployment:
```shell
# Pick one Pod that belongs to the Deployment, and view its logs
kubectl logs deployment/configmap-env-var
```

You should see an output similar to the below:
```
Found 3 pods, using pod/configmap-env-var-6d9ff89fb6-bzcf6
Thu Jan  4 16:30:35 UTC 2024 The basket is full of mangoes
Thu Jan  4 16:30:45 UTC 2024 The basket is full of mangoes
Thu Jan  4 16:30:55 UTC 2024 The basket is full of mangoes
```

This demonstrates the scenario of updating environment variables in a Pod that are derived
from a ConfigMap. Changes to the ConfigMap values are applied to the Pod during the subsequent
rollout. If Pods get created for another reason, such as scaling up the Deployment, then the new Pods
also use the latest configuration values; if you don't trigger a rollout, then you might find that your
app is running with a mix of old and new environment variable values.

> ## Update configuration via a ConfigMap in a multi-container Pod

Reviewer: ```suggestion
## Update configuration via a ConfigMap in a multi-container Pod {#rollout-configmap-multiple-containers}
```

The kubectl tool updates the ConfigMap accordingly (if you see an error, try again).

Here's an example of how that manifest could look after you edit it:

```yaml
apiVersion: v1
data:
  color: blue
kind: ConfigMap
# You can leave the existing metadata as they are.
# The values you'll see won't exactly match these.
metadata:
  creationTimestamp: "2024-01-05T08:12:05Z"
  name: color
  namespace: configmap
  resourceVersion: "1801272"
  uid: 80d33e4a-cbb4-4bc9-ba8c-544c68e425d6
```

Loop over the service URL for few seconds. 
```shell
# Cancel this when you're happy with it (Ctrl-C)
while true; do curl --connect-timeout 7.5 http://localhost:8080; sleep 10; done
```

You should see the output change as follows:
```
Fri Jan  5 08:14:00 UTC 2024 My preferred color is red
Fri Jan  5 08:14:02 UTC 2024 My preferred color is red
Fri Jan  5 08:14:20 UTC 2024 My preferred color is red
Fri Jan  5 08:14:22 UTC 2024 My preferred color is red
Fri Jan  5 08:14:32 UTC 2024 My preferred color is blue
Fri Jan  5 08:14:43 UTC 2024 My preferred color is blue
Fri Jan  5 08:15:00 UTC 2024 My preferred color is blue
```

> ## Update configuration via a ConfigMap in a Pod possessing a Sidecar Container

Reviewer: ```suggestion
## Update configuration via a ConfigMap in a Pod possessing a sidecar container {#rollout-configmap-sidecar}
```

The above scenario can be replicated by using a [Sidecar Container](/docs/concepts/workloads/pods/sidecar-containers/) as a helper container to write the HTML file. As a Sidecar Container is conceptually an Init Container, it is guaranteed to start before the main webserver container. This ensures that the HTML file is always available when the webserver is ready to serve it. Please see [Enabling sidecar containers](/docs/concepts/workloads/pods/sidecar-containers/#enabling-sidecar-containers) to utilize this feature.

> If you are continuing from the previous scenario, you can reuse the ConfigMap named `color` for this scenario. If you are executing this scenario independently, use the `kubectl create configmap` command to create a ConfigMap from [literal values](/docs/tasks/configure-pod-container/configure-pod-configmap/#create-configmaps-from-literal-values):

Reviewer: Please wrap these lines.

Author: This is addressed in the latest commit.

The kubectl tool updates the ConfigMap accordingly (if you see an error, try again).

Here's an example of how that manifest could look after you edit it:

```yaml
apiVersion: v1
data:
  color: green
kind: ConfigMap
# You can leave the existing metadata as they are.
# The values you'll see won't exactly match these.
metadata:
  creationTimestamp: "2024-02-17T12:20:30Z"
  name: color
  namespace: default
  resourceVersion: "1054"
  uid: e40bb34c-58df-4280-8bea-6ed16edccfaa
```

Loop over the service URL for few seconds.
```shell
# Cancel this when you're happy with it (Ctrl-C)
while true; do curl --connect-timeout 7.5 http://localhost:8081; sleep 10; done
```

You should see the output change as follows:
```
Sat Feb 17 13:12:35 UTC 2024 My preferred color is blue
Sat Feb 17 13:12:45 UTC 2024 My preferred color is blue
Sat Feb 17 13:12:55 UTC 2024 My preferred color is blue
Sat Feb 17 13:13:05 UTC 2024 My preferred color is blue
Sat Feb 17 13:13:15 UTC 2024 My preferred color is green
Sat Feb 17 13:13:25 UTC 2024 My preferred color is green
Sat Feb 17 13:13:35 UTC 2024 My preferred color is green
```

> ## Update configuration via a Immutable ConfigMap mounted as a Volume

Reviewer: ```suggestion
## Update configuration via an immutable ConfigMap that is mounted as a Volume {#rollout-configmap-immutable-volume}
```

> {{< warning >}}
> Immutable ConfigMaps are especially used for configuration which is constant and is not expected to change over time.   
> In a production use-case, it is advisable to not change the contents of Immutable ConfigMaps unless absolutely necessary.
> {{< /warning >}}

Reviewer: Warning is too strong
```suggestion
{{< note >}}
Immutable ConfigMaps are especially used for configuration that is constant and is **not**
expected to change over time. Marking a ConfigMap allows a performance improvement where the
kubelet does not watch for changes.

If you do need to make a change, you should plan to either:
- change the name of the ConfigMap, and switch to running Pods that reference the new name
- replace all the nodes in your cluster that have previously run a Pod that used the old value
- restart the kubelet on any node where the kubelet previously loaded the old ConfigMap
{{< /note >}}
```

  name: season
  namespace: default
  resourceVersion: "3840"
  uid: 10cdc5e6-47b8-4473-a9f3-ed401c2073f9
```

Attempt to save the changes. You should see that the save fails and a temporary manifest is opened which contains the reason for failure:
```
# reopened with the relevant failures.
#
# configmaps "season" was not valid:
# * data: Forbidden: field is immutable when `immutable` is set
#
```

Without any further change, save the temporary manifest. You should see an output similar to:
```
A copy of your changes has been stored to "/var/folders/nn/bb6y1j8d69n9lhqppjlrfthm0000gn/T/kubectl-edit-139284375.yaml"
error: Edit cancelled, no valid changes were saved.
```

Once a ConfigMap is marked as immutable, it is not possible to revert this change nor to mutate the contents of the data or the binaryData field.  
In order to modify the contents, the only option is to delete and recreate the ConfigMap.

Delete and recreate the ConfigMap `season` by using the temporary manifest as follows:
```shell
# Use the appropriate path to the temporary manifest
kubectl replace --force -f /var/folders/nn/bb6y1j8d69n9lhqppjlrfthm0000gn/T/kubectl-edit-139284375.yaml

> ```

Reviewer: This isn't always enough, because the kubelet doesn't watch for changes to immutable ConfigMaps.

Instead, I would put a date in the ConfigMap name, and teach switching to referencing a new different ConfigMap.

For example:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: "company-name-20150801"
data:
  company_name: "ACME, Inc." # existing fictional company name
immutable: true
```

and then switch to something like:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: "company-name-20240312"
data:
  company_name: "Fiktivesunternehmen GmbH" # new fictional company name
immutable: true
```

Author: This is implemented in the latest commit.

Tail (follow the latest entries in) the logs of one of the pods that belongs to this Deployment:
```shell
# As the text explains, the output does NOT change
kubectl logs deployments/immutable-configmap-volume --follow
```
Notice that the output remains **unchanged**, even though you recreated the ConfigMap:
```
Tue Mar  5 12:02:07 UTC 2024 My preferred season is summer
Tue Mar  5 12:02:17 UTC 2024 My preferred season is summer
Tue Mar  5 12:02:27 UTC 2024 My preferred season is summer
Tue Mar  5 12:02:37 UTC 2024 My preferred season is summer
```

> {{< caution >}}
> Although the ConfigMap has been recreated, the existing Pods maintain a mount point to the deleted ConfigMap. To force an update, you would need to clean up every existing reference to the old ConfigMap. In this scenario, it equates to deleting the Deployment and recreating it. Exercise caution before executing these steps in a production environment.
> 
> A simple restart of the existing Deployment using the `kubectl rollout restart` command causes the Pods to be recreated, but is insufficient to refresh the contents of the ConfigMap.
> {{< /caution >}}

Reviewer: ```suggestion
{{< note >}}
A simple restart of the existing Deployment using the `kubectl rollout restart` command causes the Pods
to be recreated, but is insufficient to refresh the contents of the ConfigMap.
{{< /note >}}

Although the ConfigMap has been recreated, any existing Pods maintain a mount point to the volume
representing the deleted ConfigMap. To force an update, you would need to clean up every existing
reference to the old ConfigMap. In this scenario, it equates to deleting the Deployment and recreating it.
You should be cautious about executing similar steps in a production environment.
```

Check the Pods:
```shell
kubectl get pods --selector=app.kubernetes.io/name=immutable-configmap-volume
```

You should see an output similar to:
```
NAME                                          READY   STATUS    RESTARTS   AGE
immutable-configmap-volume-5c677f6494-dxjwg   1/1     Running   0          31s
immutable-configmap-volume-5c677f6494-k54fc   1/1     Running   0          31s
immutable-configmap-volume-5c677f6494-w9b4c   1/1     Running   0          31s
```

View the logs for a Pod in this Deployment:
```shell
# Pick one Pod that belongs to the Deployment, and view its logs
kubectl logs deployment/immutable-configmap-volume
```

You should see an output similar to the below:
```
Found 3 pods, using pod/immutable-configmap-volume-5c677f6494-dxjwg
Tue Mar  5 12:40:43 UTC 2024 My preferred season is winter
Tue Mar  5 12:40:53 UTC 2024 My preferred season is winter
Tue Mar  5 12:41:03 UTC 2024 My preferred season is winter
```

## Summary

Changes to a ConfigMap mounted as a Volume on a Pod are available seamlessly after the subsequent kubelet sync.

Changes to a ConfigMap that configures environment variables for a Pod are available after the subsequent rollout for the Pod.

> Once a ConfigMap is marked as immutable, it is not possible to revert this change nor to mutate the contents of the data or the binaryData field. You can only delete and recreate the ConfigMap. Because existing resources maintain a mount point to the deleted ConfigMap, it is necessary to ensure that every existing reference to the old ConfigMap (not just Pods, any reference) has been cleaned up.

Reviewer: ```suggestion
Once a ConfigMap is marked as immutable, it is not possible to revert this change
(you cannot make an immutable ConfigMap mutable), and you also cannot make any change to
the contents of the `data` or the `binaryData` field. You can delete and recreate the ConfigMap,
or you can make a new different ConfigMap. When you delete a ConfigMap, running containers
and their Pods maintain a mount point to any volume that referenced that existing ConfigMap.
```

---

> reviewers:
>   - sftim

Reviewer: ```suggestion
```
I don't have any special ConfigMap skills, so I wouldn't list me here.

Author: I've handled this in the latest commit.

> title: Updating Configuration via a Configmap

Author: Thanks, I've addressed this in the latest commit.

content_type: tutorial
weight: 20
---

<!-- overview -->

> This page provides a step-by-step example of updating configuration within a Pod via a ConfigMap and builds upon the [Configure a Pod to Use a ConfigMap](/docs/tasks/configure-pod-container/configure-pod-configmap/) task. At the end of this tutorial, you will understand how to change the configuration for a running application. 

Author: Done.

* Update environment variables of a Pod via a ConfigMap
* Update configuration via a ConfigMap in a multi-container Pod
* Update configuration via a ConfigMap in a Pod possessing a Sidecar Container

<!-- lessoncontent -->

## Update configuration via a ConfigMap mounted as a Volume {#rollout-configmap-volume}

Use the `kubectl create configmap` command to create a ConfigMap from [literal values](/docs/tasks/configure-pod-container/configure-pod-configmap/#create-configmaps-from-literal-values):

```shell
kubectl create configmap sport --from-literal=sport=football
```
Below is an example of a Deployment manifest with the ConfigMap `sport` mounted as a
{{< glossary_tooltip text="volume" term_id="volume" >}} into the Pod's only container.
{{% code_sample file="deployments/deployment-with-configmap-as-volume.yaml" %}}

Create the Deployment:
```shell
kubectl apply -f https://k8s.io/examples/deployments/deployment-with-configmap-as-volume.yaml
```

Check the Deployment:
```shell
kubectl get deployment configmap-volume
```

You should see an output similar to:
```
NAME               READY   UP-TO-DATE   AVAILABLE   AGE
configmap-volume   3/3     3            3           19s

> ```

Author: Thanks, I've removed the deployment check from all relevant places. At the place where a kubectl rollout restart is performed, I've kept the deployment check as its important to verify the rollout was completed successfully.

Check the pods for this Deployment (matching by
{{< glossary_tooltip text="selector" term_id="selector" >}}):
```shell
kubectl get pods --selector=app.kubernetes.io/name=configmap-volume
```

You should see an output similar to:
```
NAME                                READY   STATUS    RESTARTS   AGE
configmap-volume-6b976dfdcf-qxvbm   1/1     Running   0          72s
configmap-volume-6b976dfdcf-skpvm   1/1     Running   0          72s
configmap-volume-6b976dfdcf-tbc6r   1/1     Running   0          72s

> ```

Author: It is important to verify that the Pods are in a ready state before proceeding with the tutorial. So I propose to keep this step.

The kubectl tool updates the ConfigMap accordingly (if you see an error, try again).

Here's an example of how that manifest could look after you edit it:

```yaml
apiVersion: v1
data:
  sport: cricket
kind: ConfigMap
# You can leave the existing metadata as they are.
# The values you'll see won't exactly match these.
metadata:
  creationTimestamp: "2024-01-04T14:05:06Z"
  name: sport
  namespace: default
  resourceVersion: "1743935"
  uid: 024ee001-fe72-487e-872e-34d6464a8a23
```

You should see the following output:
```
configmap/sport edited
```

Tail (follow the latest entries in) the logs of one of the pods that belongs to this Deployment:
```shell
kubectl logs deployments/configmap-volume --follow
```
After few seconds, you should see the log output change as follows:
```
Thu Jan  4 14:11:36 UTC 2024 My preferred sport is football
Thu Jan  4 14:11:46 UTC 2024 My preferred sport is football
Thu Jan  4 14:11:56 UTC 2024 My preferred sport is football
Thu Jan  4 14:12:06 UTC 2024 My preferred sport is cricket
Thu Jan  4 14:12:16 UTC 2024 My preferred sport is cricket
```

> When you have a ConfigMap that is mapped into a running Pod using either a `configMap` volume or a `projected` volume, and you update that ConfigMap, the running Pod sees the update almost immediately. However, your application only sees the change if it is written to either poll for changes, or watch for file updates. An application that loads its configuration once at startup will not notice a change.

Reviewer 3: Please wrap these long lines manually.

Author: Done.

# The values you'll see won't exactly match these.
metadata:
  creationTimestamp: "2024-01-04T16:04:19Z"
  name: fruits
  namespace: default
  resourceVersion: "1749472"

```

You should see the following output:
```
configmap/fruits edited
```

Tail the logs of the Deployment and observe the output for few seconds:
```shell
# As the text explains, the output does NOT change
kubectl logs deployments/configmap-env-var --follow
```

Notice that the output remains **unchanged**, even though you edited the ConfigMap:
```
Thu Jan  4 16:12:56 UTC 2024 The basket is full of apples
Thu Jan  4 16:13:06 UTC 2024 The basket is full of apples
Thu Jan  4 16:13:16 UTC 2024 The basket is full of apples
Thu Jan  4 16:13:26 UTC 2024 The basket is full of apples
```

{{< note >}}
Although the value of the key inside the ConfigMap has changed, the environment variable in the Pod still shows the earlier value. This is because environment variables for a process running inside a Pod are **not** updated when the source data changes; if you wanted to force an update, you would need to have Kubernetes replace your existing Pods. The new Pods would then run with the updated information.

> {{< /note >}}

Reviewer 3: Since this subsection only differs from the previous one on how the changes are effected, I'd suggest we focus on the differences. Even for a tutorial, we don't need to enumerate each and every steps.

Author: I believe it will help the reader to copy/paste the commands for ease of learning. So I propose to keep the steps..

Author: Thanks, I've made the necessary changes.

Reviewer 3: If you want to keep this subsection rather than drop it completely, you may want to focus on the "meaningful" steps, which is that users have to recreate their pods. There are many ways to recreate them. The simplest one is to delete and recreate.
I mean we don't need the lines on creating and verifying the ConfigMap or the Deployment.

Author: The deployment manifest in the section for ConfigMap as environment variable is different to the previous section where the ConfigMap is mounted as a Volume. So I think it is necessary to provide the exact manifest on how to create the deployment, so that the reader does not make a mistake here.
